src/compiler/crystal/codegen/primitives.cr

```diff
@@ -731,7 +731,7 @@ class Crystal::CodeGenVisitor
     allocate_aggregate base_type
 
     unless type.struct?
-      type_id_ptr = aggregate_index(@last, 0)
+      type_id_ptr = aggregate_index(llvm_struct_type(type), @last, 0)
```

Comment:
I'm quite certain that here is the problem. I looked into it because "unless struct get struct type" sounded weird. Locally I can build fine with every other place using gep2 but here. This said, I couldn't fix it and, weirdly, removing entirely this code (`unless ... end`) makes everything work. @HertzDevil if you have any clue, let me know, otherwise we call Ary.

Reply:
The `llvm_type` of any reference type is a pointer, `llvm_struct_type` is the pointee's layout:

```crystal
class Foo
  @x = ""
end

# type                   # => Foo
# llvm_struct_type(type) # => %Foo = type { i32, %String* }
# llvm_type(type)        # => %Foo*
Foo.new
```